incubator/virtualcluster/pkg/vn-agent/server/route.go

```diff
@@ -113,27 +122,110 @@ func (s *Server) InstallHandlers() {
-	klog.V(4).Infof("request %+v", req.Request.URL)
-
-	req.Request.URL.Host = s.config.KubeletServerHost
-	req.Request.URL.Scheme = "https"
+func getSuperMasterAddress() (string, error) {
+	ip := os.Getenv("KUBERNETES_SERVICE_HOST")
+	if ip == "" {
+		return "",
+			errors.New("cannot get super apiserver ip through env \"KUBERNETES_SERVICE_HOST\"")
+	}
+	httpsPort := os.Getenv("KUBERNETES_SERVICE_PORT_HTTPS")
+	if httpsPort == "" {
+		return "",
+			errors.New("cannot get super apiserver https port through env \"KUBERNETES_SERVICE_PORT_HTTPS\"")
+	}
+	return ip + ":" + httpsPort, nil
+}
 
-	// there must be a peer certificate in the tls connection
-	if req.Request.TLS == nil || len(req.Request.TLS.PeerCertificates) == 0 {
-		resp.ResponseWriter.WriteHeader(http.StatusForbidden)
+func getSATokenAndCA() (saToken, ca []byte, err error) {
+	saToken, err = ioutil.ReadFile(InClusterSATokenPath)
+	if err != nil {
 		return
 	}
-	tenantName := req.Request.TLS.PeerCertificates[0].Subject.CommonName
-	TranslatePath(req, tenantName)
-
-	klog.V(4).Infof("request after translate %+v", req.Request.URL)
+	ca, err = ioutil.ReadFile(InClusterCACrtPath)
+	if err != nil {
+		return
+	}
+	return
+}
 
-	transport := &http.Transport{
-		TLSClientConfig: &tls.Config{
-			InsecureSkipVerify: true,
-			Certificates:       []tls.Certificate{s.config.KubeletClientCert},
-		},
+func (s *Server) proxy(req *restful.Request, resp *restful.Response) {
+	klog.V(4).Infof("request %+v", req.Request.URL)
+	var transport = &http.Transport{}
+
+	if s.config.KubeletClientCert != nil {
+		klog.Info("will forward request to kubelet")
+		// forward request to kubelet
+		req.Request.URL.Host = s.config.KubeletServerHost
+		req.Request.URL.Scheme = "https"
+
+		// there must be a peer certificate in the tls connection
+		if req.Request.TLS == nil || len(req.Request.TLS.PeerCertificates) == 0 {
+			resp.ResponseWriter.WriteHeader(http.StatusForbidden)
+			return
+		}
+		tenantName := req.Request.TLS.PeerCertificates[0].Subject.CommonName
+		TranslatePath(req, tenantName)
+
+		klog.V(4).Infof("request after translate %+v", req.Request.URL)
+
+		transport = &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+				Certificates:       []tls.Certificate{*s.config.KubeletClientCert},
+			},
+		}
+	} else {
+		klog.Info("will forward request to super apiserver")
+		// forward request to super apiserver
+		if req.Request.TLS == nil {
+			resp.ResponseWriter.WriteHeader(http.StatusForbidden)
+			return
+		}
+		// 1. get super master address
+		superHttpsAddr, err := getSuperMasterAddress()
+		if err != nil {
+			klog.Errorf("fail to get apiserver address: %s", err)
+			resp.ResponseWriter.WriteHeader(http.StatusNotFound)
+			resp.ResponseWriter.Write([]byte(err.Error()))
+			return
+		}
+
+		// 2. As we use the sa of the vn-agent, to forward the request, make sure
+		// the sa has the permission to access pod
+		saToken, ca, err := getSATokenAndCA()
```

Comment:
step 1, step 2.  how about just take advantage of  `k8s.io/client-go/rest.InClusterConfig`

Reply:
remove function getSATokenAndCA and use rest.InClusterConfig instead. By default the `CAData` field of the `config.TLSClientConfig` is empty, so I read `CAFile` instead.